### ISSUE TITLE
Helm: Do not allow specifying Dagger version

### ISSUE THREAD

**Issue by gerhard** [dagger/dagger#8301]
## Initial report by @smolinari 

<img width="996" alt="image" src="https://github.com/user-attachments/assets/6db84b4d-3af5-4e06-8347-75bf8fb09c34">

## The actual problem + solution

<img width="1098" alt="image" src="https://github.com/user-attachments/assets/63fdb613-814d-47a9-b0a1-af68c1d19685">

## The better solution
Now that we publish a new Helm version whenever there is a new Dagger version - in other words the chart version and the Dagger version are moving in lock-step - we should not allow specifying different versions for Dagger.

We should still support setting custom images in case someone wants to use a specific image variant. We should add a comment warning which explains why the two versions need to be kept in lock-step.

**Comment by smolinari** [comment]
So, just for my understanding for the future, let's say I have a bunch of CI processes running on engine 0.X.X., If I upgrade the Helm chart, it could break the processes. This is because Dagger is still considered pre-production (per SemVer) and the API can have breaking changes (as you noted will happen in 0.13.0 in our Discord chat @gerhard), so upgrading the Helm chart must be done with caution and understanding which Engine versions will have breaking changes is very important. 

Once Dagger is 1.0.0, only major version changes (i.e. 2.x.x, 3.x.x, 4.x.x etc.) will need external work done to upgrade the Helm chart. 

Is that a correct take?
 
Scott

**Comment by gerhard** [comment]
Yes, that is correct.

Before `1.0`, minor versions can be considered major ones. We tend to introduce the most breaking changes in minors, e.g.:
- `v0.12.0` https://github.com/dagger/dagger/blob/main/CHANGELOG.md#v0120---2024-07-12
- `v0.11.0` https://github.com/dagger/dagger/blob/main/CHANGELOG.md#v0110---2024-04-03

There are also cases when we don't, e.g.:
- `v0.10.0` https://github.com/dagger/dagger/blob/main/CHANGELOG.md#v0100---2024-02-27

The truth is that we also introduce smaller breaking changes in patch versions (search for breaking): https://github.com/dagger/dagger/blob/main/CHANGELOG.md

In conclusion: always look at the release notes before upgrading the Helm chart version. And always pin!

cc @jedevc

**Comment by marcosnils** [comment]
> ## The better solution
> Now that we publish a new Helm version whenever there is a new Dagger version - in other words the chart version and the Dagger version are moving in lock-step - we should not allow specifying different versions for Dagger.

 @gerhard  maybe we can still support this by adding a bit of logic to our helm chart about supported verisons? Seems like using both the fail [1] and semver [2] sprig package we could prevent newer helm charts to set engine versions older than whatever semver def we want

[1]: https://masterminds.github.io/sprig/flow_control.html
[2] https://masterminds.github.io/sprig/semver.html

thoughts?

**Comment by gerhard** [comment]
That would work if the `appVersion` was semver, but it could be any image tag, including:
- `be158d163f0576d998abe399e6612381dc8ef8a8` (a specific git sha)
- `main` (meaning latest dev build)
- `main-wolfi` (a dev variant)
- `v0.12.7-gpu` (this looks like semver, but it's not)

Since the `appVersion == version`, we should instead expose the entire image ref, rather than use the `appVersion` for the default `engine.image.tag` and, potentially, allow for situations where the `engine.image.repository` value is incompatible with the `engine.image.tag` value.

I am just finishing off the PR, will submit so that we can discuss further on the proposed changes.

**Comment by jedevc** [comment]
We could consider tagging all images with semver tags going forward, similar to how we already do for cli/engines. All the groundwork is set in https://github.com/dagger/dagger/pull/7858, we'd just need to plumb it into the engine publishing logic.

*Probably* out of scope for v0.13.0, but maybe to consider for later.

**Comment by gerhard** [comment]
Yes, I remember us talking about that. When we do that, we should definitely revisit this.
